feature_engine/dataframe_checks.py

```diff
@@ -129,3 +129,41 @@ def _check_contains_inf(X: pd.DataFrame, variables: List[Union[str, int]]) -> No
             "Some of the variables to transform contain inf values. Check and "
             "remove those before using this transformer."
         )
+
+
+def _check_X_y_pd_np_mismatch(
+    X: Union[pd.DataFrame, np.ndarray],
+    y: Union[pd.Series, np.ndarray],
+):
+    """
+    Handles 3 cases
+    1. X is an ndarray and y is a Series - converts X to DataFrame with y's index
+    2. X is a DataFrame and y is an ndarray - converts y to Series with X's index
+    3. X is a DataFrame and y is a Series, but their indexes don't match
+    - raises an error
```

Comment:
This check is looking really good. 

I would refactor it slightly so that it returns X and y as a Dataframe and Series, so we do not have additional work to do within the classes. So:

```
def _check_X_y_pd_np_mismatch(
    X: Union[pd.DataFrame, np.ndarray],
    y: Union[pd.Series, np.ndarray],
) -> Tuple[pd.DataFrame, pd.Series]:
```

The only missing functionality would be for when both X and y are arrays, that instead of being returned unchanged, we should return a df and a series.